rfcs/20201014-VTune-ITT-tagging/README.md

```diff
@@ -0,0 +1,142 @@
+![](images/timeline.JPG "oneDNN primitives timeline chart")
+
+### 2. get platform information such as L1/L2 cache miss or level of FP vectorization on primitive level   
+For below diagram, users can group profiling results by Task Types, and then VTune will group information by
+oneDNN primitives tagged as different Task Types.   
+Therefore, users can see platform level detail information like FP vectorization on primitives level.  
+Users can understand performance issues among each oneDNN primitives.  
+![](images/group.JPG "group by oneDNN primitives")
+
+### 3. map primitive with related computation kernels.  
+By tagging primitives, users can also map primitive with computation kernels.  
+For example, users can understand one inner primitive contains several gemm_nocopy_f32 kernels.  
+![](images/ops_mapping.JPG "mapping between oneDNN primitives and kernels")
+
+## Proposal
+
+Add VTune ITT tasks for all primitives.
+Users can see profiling results grouped by primitives. 
+primitive grouping is more useful than function grouping to DL developers.
+### Implementation
+To enable VTune ITT tagging, we includes ittnotify related source files in projects https://github.com/intel/ittapi under src/common/ittnofiy folder.  
+We also create ittnotify.cpp and ittnotify.hpp under /src/common folder to use ITT functions from copied ITT source files.
+We introduce a new namespace itt under dnnl namespace.
+In the same sapce we have 4 DNNL_API for using ITT tagging.
+
+| functions  | Note |  
+|---------------------|---------------------|  
+| void DNNL_API primitive_task_start(primitive_kind_t kind) |  use this function to tag start of a ITT task with primitive kind info  | 
+| void DNNL_API primitive_task_end()  |    use this function to tag end of a ITT task    | 
+| primitive_kind_t DNNL_API primitive_task_get_current_kind() |   get the primitive task for current ITT task       | 
+| bool DNNL_API get_itt(__itt_task_level level) |    get ITT level.  enable different ITT tagging by using this level variable  |   
+
+We also define a enum type with different itt task level.  
+Here is a table for meaning of all itt task levels.  
+
+| __itt_task_level  | Note |  
+|---------------------|---------------------|  
+| __itt_task_level_none |   no ITT event will be triggered  | 
+| __itt_task_level_low  |    ITT events are only triggered in master thread     | 
+| __itt_task_level_high |    ITT events are triggered in all OMP/TBB threads      |   
+
+There is a new struct itt_task_level_t with only one int variable in it.  
+~~~cpp
+struct itt_task_level_t {
+    int level;
+};
+~~~
+
+#### Implementation details :  tag primitive execution with ITT
+---
+Here are pesudo codes for how we define primitive execution on master thread as a ITT task.
+~~~cpp
+status_t dnnl_primitive_execute(const primitive_iface_t *primitive_iface,
+    ....
+    if (itt::get_itt(itt::__itt_task_level_low))
```

Comment:
Is the suggestion to change the oneDNN source code to enable primitive tagging, or is the tagging performed at a higher level (read -- framework integration)?

Given that the proposal provides APIs for tagging primitives, this example becomes confusing.

Reply:
yes. we did the changes in oneDNN to enable primitive tagging.
here is the MR:
https://gitlab.devtools.intel.com/ipl/dl/mkl-dnn/-/merge_requests/2091

no need for highlevel FW tagging.

We could remove the example, because it is not a general example. it is only for primitive tagging.

Reply:
I was mostly referring to the proposal to introduce some `DNNL_API` functions. Since primitive tagging will be done inside the library and not externally, I don't see why these should be exported using `DNNL_API`. Hence I suggest to drop this attribute.

Reply:
Hi 
If I removed DNNL_API, benchdnn, tests, and other samples will failed at linking stage due to "undefined reference to `dnnl::impl::itt::primitive_task_start". 
Therefore, for those standalone application, those 4 functions  might still need to be exported as DNNL_API functions.

Reply:
Is it necessary to use these functions in benchdnn? I thought these will be used only in the main source code.

Reply:
people indeed use benchdnn to validate different computation kernel among architectures like ATS, DG1.
The feature should also help to identify different detail platform info like GPU memory usage among different primitives.

Reply:
If I understand correctly, these functions should be used in oneDNN source code (not benchdnn), specifically in `status_t dnnl_primitive_execute`. The question is why are these functions necessary in benchdnn?

Reply:
the linking is coming from below file.
`src/common/dnnl_thread_parallel_nd.hpp`

Both benchdnn or gtest uses below template and prototype codes from dnnl_thread_parallel_nd.hpp, and primitve_task_start is called inside this template.
```
template <typename F>
void parallel(int nthr, F f) {
...
        if (ithr_ && itt_enable)
            itt::primitive_task_start(
                    task_primitive_kind);

```

If we don't export those functions as DNN_API, both benchdnn and gtest  failed at dnnl::impl::parallel function at linking stage.

Hope the explanation make sure to you.

Reply:
Thanks for the explanation, now I understand the problem. It seems that `dnnl_thread_parallel_nd.hpp` is included only because of threadpool testing. I think we may need to come up with some solution for it later. Resolving the discussion.

Reply:
Also, please include this code snippet into the RFC.

Reply:
@louie-tsai Seems like this issue can be resolved by including necessary headers into src/common/dnnl_thread_parallel_nd.hpp to make this functions visible during benchdnn compilation. Could you please try this approach?